Call TypeDescriptor.GetProperties before calling Refresh in ReflectionCachesUpdateHandler

stephentoub opened the pull request:
Try to fix https://github.com/dotnet/runtime/issues/51588

According to the docs:
"Before you make a call to the Refresh method to clear the cache, you need to call the GetProperties method for the specific assembly to cache the information first."

Counterintuitive, but ok.

cc: @jeffhandley, @safern

stephentoub committed:
Call TypeDescriptor.GetProperties before calling Refresh in ReflectionCachesUpdateHandler

safern reviewed (approved):
Hmm interesting 😕

jeffhandley reviewed (approved):
Thanks! Any idea why it was only failing on macOS?

stephentoub commented:
> Thanks! Any idea why it was only failing on macOS?

My guess is the failures are happening when this test runs concurrently with some other test, and without the GetProperties call, Refresh ends up not being thread-safe.  At least that's my very limited understanding from the docs and a quick perusal of the source.  If that's true, then this isn't limited to just macOS, and that's simply the only place we've seen it so far.

jeffhandley commented:
Seems like a reasonable hunch. Thanks for addressing it so swiftly!

danmoseley commented:
Looks like it's also failed in Linux now so not Mac specific

stephentoub commented:
And also not fixed by this change.  It's possible TypeDescriptor.Refresh simply isn't thread-safe, violating rules around statics in .NET.  I'll need to look deeper.

stephentoub committed:
Isolate ReflectionCachesUpdateHandlerTests

TypeDescriptor.Refresh isn't thread-safe, trying to enumerate a hashtable that's concurrently mutated, leading to a concurrent modification exceptions.  We'll need to decide if we're ok with the limitation or if we need to do surgery on TypeDescriptor to fix it.
